### PR TITLE
logmatcher: fix storing matches if source handle changes

### PR DESCRIPTION
**Pull request by alltilla**
Example config for reproduction:
```
@version: 4.5
@include "scl.conf"

source s_local {
  example-msg-generator(template("foo bar baz"));
};

destination d_local {
  stdout(template("$A_foobar $MESSAGE $Z_barbaz"));
};

log {
  source(s_local);

  parser {
    regexp-parser(
      patterns("(?<A_foobar>[a-z]+) (?<MESSAGE>[a-z]+) (?<Z_barbaz>[a-z]+)")
    );
  };

  destination(d_local);
};
```

### CODE REVIEW

File: lib/logmatcher.c [hunk 1]
```diff
@@ -416,6 +418,7 @@ log_matcher_pcre_re_feed_value(LogMatcherPcreRe *self, LogMessage *msg,
        * make it less appearent, this is not always the case: it does not
        * happen if the new value does not fit the old NVEntry).  */
       log_matcher_pcre_re_save_source_value_to_avoid_clobbering(result);
+      result->source_handles_value_changed = TRUE;
```
Comment (bazsi) [inline lib/logmatcher.c:421]: is it enough that we set this to TRUE at this point? I understand that name-value pairs that are set _prior_ to MESSAGE will be indirect values, which will be resolved once MESSAGE is assigned a new value. But I think we could determine that our input is rewritten by the regexp and in those cases we should never use indirect values.

Comment (alltilla) [reply lib/logmatcher.c]: Changed the implementation to check it beforehand.

**Comment by MrAnno** [comment]
Nice catch!

File: lib/logmatcher.c [hunk 2]
```diff
@@ -384,8 +424,26 @@ typedef struct _LogMatcherPcreMatchResult
...
 
+static void
+log_matcher_pcre_match_result_init(LogMatcherPcreMatchResult *self, LogMatcherPcreRe *matcher,
+                                   const gchar *source_value, gssize source_value_len, NVHandle source_handle)
+{
+  self->match_data = pcre2_match_data_create_from_pattern(matcher->pattern, NULL);
+  self->source_value = source_value;
+  self->source_value_len = source_value_len != -1 ? source_value_len : strlen(source_value);
+  self->source_handle = source_handle;
+  self->force_direct_set = _is_input_overwritten_by_named_capture(matcher, source_handle);
```
Comment (bazsi) [inline lib/logmatcher.c:438]: I think this function does not need to be called from every match operation. It's enough to call it just once, during initialization. Doing it for every match seems to be a tad expensive to me.

Comment (alltilla) [reply lib/logmatcher.c:438]: The problem is that the `value_handle` only comes with `log_matcher_match()` or `log_matcher_match_value()`. Compile time we do not know which value will be used for matching, or a not a value at all (buffer, template).

I have checked the code, and we always call `match()` or `match_value()` with the same value handle on the call sites, so I can store the value handle in init, but it seems strange to me.

Maybe we could cache the last matched value handle and its overwrite behavior, and only recalculate the overwrite behavior if the cached value handle is different than the currently matched one.

What do you think?

Comment (bazsi) [reply lib/logmatcher.c]: we've discussed this IRL and concluded that we should go back to the initial solution, e.g. we would stop producing indirect nv-pairs as soon as we discover that we are overwriting the original value.

Those that are already saved as indirect matches will be copied automatically by NVTable.

**Review by bazsi — APPROVED** [review]

**Pull request merged**